### ISSUE TITLE
packaging externals

### ISSUE THREAD

**Issue by radarsat1** [siconos/siconos#95]
Probably the last thing I'll have to do regarding the deb packages is figure out what to do about siconos-externals, which probably should not exist.  I mean, it's fine to leave in the git repository for easy access, but for packaging, probably individual components should get their own packages if they don't already have one.

I know there is some history to the use of each of the libraries within Siconos, but from my naive perspective, I thought it might be useful to go through it a piece at a time and discuss what we could do.

* blas_lapack
  * cblas_header/cblas.h
  isn't this part of the CBLAS package from netlib.org, shouldn't it go under netlib?
  * matrix_la.h
  I don't find any calls in Siconos to the functions defined here.
  * Siconos*.h
  These could be moved back into numerics safely, right?

* netlib
  * dctemplates, dftemplates, odepack
  These could be split off into their own Debian packages.
  odepack has been packaged by Drumwright: http://www-robotics.usc.edu/~drumwrig/code.html

* boost_contribs
  * bindings_utils.cpp
  move back into numerics?
  * cholesky.hpp
  Seems to be from http://www.guwi17.de/ublas/examples/
  Odd that it has an LGPL copyright seeing as it is header-only.  Not
  sure what that implies.
  In any case it has only one reference at:
  https://github.com/siconos/siconos/blob/master/kernel/src/utils/SiconosAlgebra/SimpleMatrixSolvers.cpp#L73
  which is followed by a comment that it should be replaced by a more robust implementation.
  * random.hpp
  Seems to be needed only by uBLAS bindings.
  * determinant.hpp, bindings_utils.hpp
  Ours. Move back to numerics?
  * expm.hpp
  Called only from kernel AlgebraTools.cpp, make it private there?
  * boost/serialization
  all these files seem to be available in latest Ubuntu boost-serialization package.  At least, I see them with `ls /usr/include/boost/serialization/` on Ubuntu 16.04.
  * optim_misc
  I think all of these are pulled from scilab.  They seem to be used strictly by some solvers in siconos-numerics.

* numeric\_bindings
  This appears to be a direct copy of https://svn.boost.org/svn/boost/sandbox/numeric_bindings/
  I'm surprised to see that it appears to have no Debian package, however some sleuthing turned up a package in Debian Science that was proposed in 2009!! https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=536270
  It is apparently abandoned, maybe we can re-open it.

* PATH_SDK
  Apparently again no Debian package.  Can't even find any references to previous attempts.  Probably a roadblock is the weird http://pages.cs.wisc.edu/~ferris/path/LICENSE, I doubt this would be accepted by Debian.
  So I realize it's an optional dependency, and we only have the headers from the library; perhaps I can remove it only from the Siconos Debian package.

* SOL
  From http://web.stanford.edu/group/SOL/software/lumod/ I think.
  Apparently no license.  Similar to PATH, should I remove from Debian package?  It's called from the LCP solver.  Alternatively we could treat it as a numerics external.  Not sure what to do to be honest.  The lack of license implies all rights reserve copyright in my understanding.  I see the source is slightly modified for Siconos.

* SuiteSparse
  This one actually does have a Debian package, in Ubuntu and Debian.  According to the README, I guess this is a modified version.  Perhaps it should be put into numerics. But how much work would it be to just use the package version?

* hairer
  I guess this is just a bunch of sources from http://www.unige.ch/~hairer/software.html
  Permissive license. Perhaps it should have its own package.

* swig, tools
  Our code, move back to numerics

Okay, after going through all that, I think I can categorise the parts into 3 types:

1. Some things should have their own package.
2. Some things should be just removed from the Siconos Debian package.
3. Some things could be made internal to numerics or kernel.

I would propose for (3) to make numerics/external and kernel/external.  Although many of them are just headers used during building of Siconos and therefore don't have an associated binary package and could be kept where they are, so maybe no change needed for those.  But if we ever wanted to get rid of the externals "component" and make things private to kernel or numerics that would be one way.

For (2) I'll just take them out of the source package with a patch if they are not going to be used.  This really only applies to PATH and *maybe* SOL?

For (1) things get complicated, as it's potentially a lot of work to make and maintain several numerical packages that are not part of Siconos.  I'm willing to do some quick packages but I don't know how complicated it could get in terms of getting them accepted to Debian.  An alternative approach could be to treat them as internal libraries to siconos-numerics.  In any case we could put them in the PPA in the meantime and see what reaction we get from Debian.  Perhaps they would be very welcome by the Debian Science team, I don't know.  Probably if we make packages they should be based on the upstream tarballs they came from instead of Siconos.

**Comment by vacary** [comment]
> netlib
> dctemplates, dftemplates, odepack
> These could be split off into their own Debian packages.
> odepack has been packaged by Drumwright: http://www-robotics.usc.edu/~drumwrig/code.html

At least for netlib, we can remove dctemplate and dftemplate that are not used. Perhaps, it is time to plug sundials  [issue #23](https://github.com/siconos/siconos/issues/23. I have a look to that it seems that it is not a huge task. sundials is maintained and the modern version of ode pack. Furthermore, it is distributed as a debian packages

**Comment by radarsat1** [comment]
Cool, might be interesting to look at.  Just to test the idea of moving externals into numerics/kernel, I made a branch:

https://github.com/siconos/siconos/tree/internalize-externals

I basically just moved externals into numerics/externals and one file (cholesky.hpp) into kernel/externals, and adjusted the CMakeList.txt files to match.  Everything seems to build fine, but with one less library.  If the externals used by numerics can be considered "internal" to numerics, then there is no use for siconos-externals.so!  Uploaded for discussion purposes.

I still suggest removing whatever we can package separately.  (Eventually... I am thinking long term here.)

**Comment by radarsat1** [comment]
Seems I spoke too soon!  Some `undefined reference` problems with externals functions when compiling numerics with tests enabled.  In any case, it's for discussion.. ;)

**Comment by radarsat1** [comment]
Just thought of another idea, I'll just write it down and test it when I get a chance: instead of moving files around, just compile externals as a static library. The only part that presents a potential issue is cholesky.hpp with its LGPL license. That might have to be compiled as a private shared library?  Although if that's the case then maybe all of siconos_externals could be made into a private shared library.

(That means, it is part of siconos-numerics package but is kept as a separate so file with a directory for that package -- discouraged, but one possible solution.)

**Comment by radarsat1** [comment]
Indeed it seems that building libsiconos-externals.a works perfectly!  Now I am able to link it statically into libsiconos-numerics and avoid making a separate package for it.  I will look into what can be removed from it and packaged externally, but perhaps first I will consult with the Debian Science mailing list.  I think it makes sense to statically link it into numerics if it consists mostly of just binding code and headers.

**Comment by vacary** [comment]
Cool. What is the process to be included in Debian scientific repo or Debian Sid ?

For the externalization, the big part  is numerics-bindings. the routine Cholesky.hpp is not really needed. We can use a boost bindings on a sparse matrix library or connect directly the csparse solver.

**Comment by radarsat1** [comment]
For Debian Science, in fact I am just getting into joining the mailing list and figuring out how to propose the package. It is a bit of a process -- I think I have to open a bug on the Debian bug tracker, and then post a mail proposing the package and find someone to sponsor it.  And I am sure they will have comments.  So.. work in progress.  But it will be worth it.  So far I have made the package based on the 4.0.0 release but I want to get as much feedback as possible on that work, so that any changes we need "upstream" (i.e. in Siconos) to support packaging can be done for 4.1.0.  In fact Debian encourages very good practices like using semantic "soname" versioning properly, etc, so it is a good exercise regardless.

Yep, cholesky is the only LGPL routine -- again, weird that it uses that license since it is header-only -- to respect the LGPL we would have to add a cpp and turn it into a library of sorts, which seems a bit silly.  in any case, it seems numerics_bindings is header-only as well, so it only makes sense to statically link it to numerics since that is what happens at compile time anyways.  I'm afraid SOL/lumod will have some issues due to the lack of license, perhaps we'll have to communicate with the authors.  In any case I'll bring it up on the Debian list.

**Comment by radarsat1** [comment]
https://lists.debian.org/debian-science/2017/01/msg00011.html

**Comment by radarsat1** [comment]
I've sent an email to the author of LUMOD to ask about the license situation.  On examination I saw that the version on the website includes LGPL-licensed myblas.c, but that we removed it in siconos-externals.  This is probably fine but unfortunately it doesn't automatically convert lumod to the public domain, we still need the author to attribute a permissive license to the part that remains.  On the other hand it probably means that it will be at *least* considered LGPL, meaning we can package it separately and link to it.  I'll keep this issue posted on whether or not I get a reply.

**Comment by xhub** [comment]
First of all thanks for making this packaging happening. It's very nice to see someone driving this projet.

I'm in favor of keeping the code that we import in a separate directory. It is easier to see where it comes from and which license is it under. I fell like it is also easier for us to advertise the siconos code as Apache2. Splitting the 2 into numerics/externals and kernel/externals is ok for me.
Especially since the boost stuff is header-only and is not in libsiconos-external.so

I'm not in favor in relying on distro pkgs for the code in siconos-externals for the following reasons:
- If we decide to remove something, we should make sure that it is easy to get under many linux distro (not only debian-based, but I would say at least redhad-based. Arch, Gentoo, slack would be a plus). There is also Mac OS and windows (ok the windows build is not working right now, but it wasn't too painful to get many , where it may be difficult to get compiled libraries for some of the code we are talking about). 
- A workaround is to have to switch to use an external lib or to compile the code in externals. However, it increases the maintenance burden since we have more packages to maintain. Also if there are some changes between the code in externals and the lib provided by the distro, it is more difficult to make the code compatible with both.
- is there a need for not installing libsiconos-externals.so?
- I don't get the rational for a static lib vs dynamic. It is also a pain on some system to deal with those.

I'll try to reply to some comments:
- HEM should be disable when distributing it because of MA28 that requires the user to agree to the terms. We could build the fortran files in a separate package and try to dynamically load the hem5 function. This way we could specify the license for this package or let the user compile the code (and comply with the license).
- LUMOD: according to http://web.stanford.edu/group/SOL/software.html, it is license under the MIT. In any case, we could disable it, it doesn't work well ...
- PATH_SDK: those are header-only. Yep it is not licensed, but my current mentor has been supervising the PATH development so I can ask him to clarify that. But I'm confident that it's ok. Also we don't package it since we require to link against it. This may change however (dynamic loading).
I'll suggest licensing those files under MIT. But we don't need to install those files. We could remove those from the debian source pkg.
- SuiteSparse: we have some extensive changes if I remember. There is also CXsparse that is in Suitesparse. Which one is packaged? Also we should check that this package is easily available.
- cholesky.hpp: I don't think it is reasonable to rely on that code for any sound computations. We should use an external library to do that. Ideally, we should use the routines in Numerics to solve linear system, to have only one code for solving linear systems of equations.


I was trying to have something like that in the past, that was the topic of Issue #33.

**Comment by radarsat1** [comment]
My rationale for "removing" externals (note, I'm only talking about removing it from the Debian packages, not from github), is that I think it doesn't really make sense to have a package containing a library that has no Siconos code it in, and that is only used by one part of Siconos anyway.  The code in siconos-externals is one of two categories:

1. It embeds code from external code bases for ease of use, sometimes modified to normalize use of BLAS, etc.
2. It contains headers or binding code that is used exclusively by or gets linked into siconos-numerics anyway.  I see no reason for these to result in a separate .so file.

The former really are separate bits of code and they should be packaged as such (imho, but I think this is also the attitude of DDs).  Since they basically don't ever change (last releases in like 1995..), I think once they are packaged properly and are made available in a more useful format, not much more work would need to be done.  Now... *that said*, I've been messing with adding headers to some of these Fortran programs and I can see that it's not always completely straight-forward, so maybe it is more of a long-term project.  I can see why it was only done for HEM5 for now.

* On the other hand the copyright issue regarding MA28 is interesting, thanks for pointing it out, I definitely missed that.  I looked up the MA28 function from HSL, and indeed the license makes it useless.  So in that case it's something that should just be removed from the Debian version of Siconos.  Too bad, but one less thing to package I guess.

* Thanks for pointing out the LUMOD licensed, I really searched the site but somehow missed that.  I think it's still worth trying to convince them to embed the license in the source or tar files.  Every Debian package requires a copyright file, and I don't think I'll get very good feedback if the license is added in a quilt patch on top of the sources.

* The PATH SDK does seem to be licensed, it just uses a really weird non-free license.  It seems pretty explicitly done to limit distribution.  Any clarification you can get would be great ;)

* If the changes to SuiteSparse are significant, I wonder if they could be useful beyond Siconos somehow?  I'll look at the changes.

I'll just note here, since it occurred to me, that as far as Debian is concerned any LGPL code is fine, since we are distributing in source code form.  So nothing special to do with cholesky.hpp in either case.

**Comment by xhub** [comment]
> My rationale for "removing" externals (note, I'm only talking about removing it from the Debian packages, not from github), is that I think it doesn't really make sense to have a package containing a library that has no Siconos code it in, and that is only used by one part of Siconos anyway.  The code in siconos-externals is one of two categories:

Ok thanks for clarifying that you only want to remove it from the debian src pkg.

> 1. It embeds code from external code bases for ease of use, sometimes modified to normalize use of BLAS, etc.
> 2. It contains headers or binding code that is used exclusively by or gets linked into siconos-numerics anyway.  I see no reason for these to result in a separate .so file.
> 
> * On the other hand the copyright issue regarding MA28 is interesting, thanks for pointing it out, I definitely missed that.  I looked up the MA28 function from HSL, and indeed the license makes it useless.  So in that case it's something that should just be removed from the Debian version of Siconos.  Too bad, but one less thing to package I guess.

Note that the version from hem5 is quite old and the license agreement is in linsp.f may be quite different from the actual one.

> * The PATH SDK does seem to be licensed, it just uses a really weird non-free license.  It seems pretty explicitly done to limit distribution.  Any clarification you can get would be great ;)

I'm pretty sure the people who wrote the code did not care much about that.

> * If the changes to SuiteSparse are significant, I wonder if they could be useful beyond Siconos somehow?  I'll look at the changes.

I don't remember all of those and I fail to keep track of it. But maybe using CXsparse would be the way to go? It looks like it supports both int32 and int64. But it requires quite some work on our end.

**Comment by vacary** [comment]
Concerning HEM5 it has to be only optional, because the integration of HEM5 in Siconos is far to complete and it is only an alternative to Lsodar for mechanical systems.

I suppose we can compile kernel without HEM5 quite easily.

We need to plug sundials it is more important.

**Comment by xhub** [comment]
Any updates on this? Did the LUMOD people replied?

**Comment by radarsat1** [comment]
I did get a reply from Prof. Saunders regarding the license, he said he was interested in making it easier but has not since changed anything in the lumod-c.zip download.  However, on the Debian Science mailing list they told me it would be okay to claim that license information was found on the download website.  I did notice also that the zip of the F77 version of LUMOD contains a CPL license file.
